File: java/ch17-02-plus/src/test/java/com/thesoftwaregorilla/tdd/money/SumTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.util.function.BiFunction;

import static org.junit.jupiter.api.Assertions.*;

//<editor-fold desc="TO DO List">
//</editor-fold>

@DisplayName("Sum Class Tests")
class SumTest {

    @DisplayName("reduce same currency")
    @ParameterizedTest(name = "{0} : {1}, {2} : {3}, Expected: {4}")
    @CsvSource({
            "USD, 5, USD, 5, 10",
            "CHF, 5, CHF, 5, 10",
            "ZAR, 5, ZAR, 5, 10"
    })
    void testReduceSameCurrency(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        testReduce(fromCurrency, fromAmount, toCurrency, toAmount, expectedAmount);
    }

    @DisplayName("reduce different currency")
    @ParameterizedTest(name = "{0} : {1}, {2} : {3}, Expected: {4}")
    @CsvSource({
            "CHF, 10, USD, 5, 10",
            "ZAR, 20, CHF, 5, 6",
            "ZAR, 34, USD, 5, 7"
    })
    void testReduceDiffCurrency(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        testReduce(fromCurrency, fromAmount, toCurrency, toAmount, expectedAmount);
    }

    private void testReduce(String fromCurrency, int fromAmount, String toCurrency, int toAmount, int expectedAmount) {
        Sum sum = new Sum(MoneyTest.getCurrencyFactory(fromCurrency).apply(fromAmount),
                MoneyTest.getCurrencyFactory(toCurrency).apply(toAmount));
        Money reduced = sum.reduce(BankTest.getBankWithRates(), toCurrency);
        assertEquals(MoneyTest.getCurrencyFactory(toCurrency).apply(expectedAmount), reduced);
    }

    @DisplayName("plus returning a sum (testPlusReturnsSum method in TDD book was part of MoneyTest)")
    @ParameterizedTest(name = "{0} : {1}")
    @CsvSource({
            "USD, CHF, 5",
            "CHF, CHF, 5",
            "ZAR, USD, 5"
    })
    public void testPlusReturnsSum(String baseCurrency, String targetCurrency, int amount) {
        Money baseMoney = MoneyTest.getCurrencyFactory(baseCurrency).apply(amount);
        Money targetMoney = MoneyTest.getCurrencyFactory(targetCurrency).apply(amount);
        Expression result = baseMoney.plus(targetMoney);
        if(baseMoney.getCurrency().equals(targetMoney.getCurrency())) {
            assertInstanceOf(Money.class, result);
            Money money = (Money) result;
            assertEquals(baseMoney.getAmount() + targetMoney.getAmount(), money.getAmount());
            assertEquals(baseMoney.getCurrency(), money.getCurrency());
        } else {
            assertInstanceOf(Sum.class, result);
            Sum sum = (Sum) result;
            assertEquals(baseMoney, sum.augend);
            assertEquals(targetMoney, sum.addend);
        }
    }


    @DisplayName("plus")
    @ParameterizedTest(name = "{0} : {1} + {2}")
    @CsvSource({
            "USD, 5, 4",
            "CHF, 7, 9",
            "ZAR, 10, 23"
    })
    void testPlus(String currency, int amount, int extraAmount) {
        final int expectedAmount = (amount * 2) + extraAmount;
        testArithmetic(currency, amount, extraAmount, expectedAmount, (money, amt) -> money.plus(MoneyTest.getCurrencyFactory(currency).apply(amt)) );
    }

    @DisplayName("times")
    @ParameterizedTest(name = "{0} : {1}, Multiplier: {2}")
    @CsvSource({
            "USD, 5, 3",
            "CHF, 7, 5",
            "ZAR, 10, 30"
    })
    void testTimes(String currency, int amount, int multiplier) {
        final int expectedAmount = (amount + amount) * multiplier;
        testArithmetic(currency, amount, multiplier, expectedAmount, Expression::times);
    }

    private void testArithmetic(String currency, int amount, int extraAmount, int expectedAmount, BiFunction<Expression, Integer, Expression> operation) {
        Expression money = MoneyTest.getCurrencyFactory(currency).apply(amount);
        Sum sum = new Sum(money, money);
        Expression result = operation.apply(sum, extraAmount);
        assertEquals(MoneyTest.getCurrencyFactory(currency).apply(expectedAmount), result.reduce(BankTest.getBankWithRates(), currency));
    }
}

File: cs/ch17-04-arithmetic/Money.Tests/SumTest.cs
using System.Runtime.ConstrainedExecution;
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using NUnit.Framework;
using TheSoftwareGorilla.TDD.Money;

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion

[TestFixture]
[Description("Sum Tests")]
public class SumTest
{
    [OneTimeSetUp]
    public static void OneTimeSetUp()
    {
        if (Bank.DefaultBank == null || Bank.DefaultBank.RateCount == 0) {
            Bank.DefaultBank = BankTest.GetBankWithRates();
        }
    }

    [TestCase("USD", 5, "USD", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("CHF", 5, "CHF", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("ZAR", 5, "ZAR", 5, 10, TestName = "reduce same currency USD 5")]
    [TestCase("CHF", 10, "USD", 5, 10, TestName = "reduce different currency CHF 10 with USD 5, expected 10")]
    [TestCase("ZAR", 20, "CHF", 5, 6, TestName = "reduce different currency ZAR 20 with CHF 5, expected 6")]
    [TestCase("ZAR", 34, "USD", 5, 7, TestName = "reduce different currency ZAR 34 with USD 5, expected 7")]
    [Category("reduction")]
    public void TestReduce(string fromCurrency, decimal fromAmount, string toCurrency, decimal toAmount, decimal expected)
    {
        var sum = new Sum(Money.For(fromAmount, fromCurrency), Money.For(toAmount, toCurrency));
        var result = sum.Reduce(BankTest.GetBankWithRates(), toCurrency);
        Assert.That(result, Is.EqualTo(Money.For(expected, toCurrency)));
    }


    [TestCase("USD", "CHF", 5, TestName = "plus returns sum USD 5")]
    [TestCase("CHF", "CHF", 5, TestName = "plus returns sum CHF 5")]
    [TestCase("ZAR", "USD", 5, TestName = "plus returns sum ZAR 5")]
    [Category("arithmetic")]
    public void TestPlusReturnsSum(string baseCurrency, string targetCurrency, decimal amount)
    {
        var baseMoney = Money.For(amount, baseCurrency);
        var targetMoney = Money.For(amount, targetCurrency);
        var result = baseMoney.Plus(targetMoney);
        if (baseMoney.Currency == targetMoney.Currency)
        {
            Assert.IsInstanceOf<Money>(result);
            Assert.That(result, Is.EqualTo(Money.For(amount + amount, baseCurrency)));
        } 
        else
        { 
            Assert.IsInstanceOf<Sum>(result);
            Sum sum = (Sum) result;
            Assert.That(sum.Augend, Is.EqualTo(baseMoney));
            Assert.That(sum.Addend, Is.EqualTo(targetMoney));
        }
    }

    [TestCase("USD", 5, 4, TestName = "plus with new Sum USD 5, 4")]
    [TestCase("CHF", 7, 9, TestName = "plus with new Sum CHF 7, 9")]
    [TestCase("ZAR", 10, 23, TestName = "plus with new Sum ZAR 10, 23")]
    [Category("arithmetic")]
    public void TestPlus(string currency, decimal amount, decimal extraAmount)
    {
        decimal expected = (amount * 2) + extraAmount;
        testArithmetic(currency, amount, extraAmount, expected, (sum, extraAmount) => sum.Plus(Money.For(extraAmount, currency)));
    }

    [TestCase("USD", 5, 3, TestName = "times with new Sum USD 5")]
    [TestCase("CHF", 7, 5, TestName = "times with new Sum CHF 7")]
    [TestCase("ZAR", 10, 30, TestName = "times with new Sum ZAR 10")]
    [Category("arithmetic")]
    public void TestTimes(string currency, decimal amount, decimal multiplier)
    {
        decimal expected = (amount + amount) * multiplier;
        testArithmetic(currency, amount, multiplier, expected, (sum, extraAmount) => sum.Times(extraAmount));
    }

    private static void testArithmetic(string currency, decimal amount, decimal extraAmount, decimal expectedAmount, Func<Expression, decimal, Expression> operation)
    {
        var money = Money.For(amount, currency);
        Sum sum = new Sum(money, money);
        var result = operation(sum, extraAmount);
        Assert.That(result.Reduce(BankTest.GetBankWithRates(), currency), Is.EqualTo(Money.For(expectedAmount,currency)));
    }
    
}